### DISCUSSION TITLE
confusion in which stream to call for RP2040

### DISCUSSION THREAD

**Discussion by SnazzyBlanket** [pschatzmann/pico-arduino#20] [General]
yesterday I got sound working without issue (and no compilation issues either )using an esp32. however I also want to use the audio tools library with the rp2040 but I must admit I am confused as there are various documents (some older some newer) which seem to recommend different ways of calling the driver.

I want to use the pcm5102 with the rp2040.

the code I got to work on the esp32 was this`/**
 * @file streams-generator-analog.ino
 * @author Phil Schatzmann
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-analog/README.md 
 * @author Phil Schatzmann
 * @copyright GPLv3
 **/
 #include <Arduino.h>

#include "WiFi.h"

#include "AudioTools.h"
#include "AudioLibs/AudioSTK.h"
#include "AudioLibs/AudioBoardStream.h"

typedef int16_t sound_t;                                   // sound will be represented as int16_t (with 2 bytes)
AudioInfo info(44100,1, 16); 
Sitar instrument(10);
STKStream<Instrmnt> in(instrument);

AnalogAudioStream out; 
StreamCopy copier(out, in);                             // copies sound into i2s

// Arduino Setup
void setup(void) {  
  // Serial.begin(115200);
    AudioLogger::instance().begin(Serial, AudioLogger::Warning);
    
    // start Output
    //Serial.println("starting Analog Output...");
    out.begin(out.defaultConfig());

    // start STK input with default configuration
   // Serial.println("starting STK...");
    auto cfgSTK = in.defaultConfig();
    cfgSTK.channels = 1;
    in.addNotifyAudioChange(out);
    in.begin(cfgSTK);
   // instrument.noteOn(440, 0.5);
   
}

// Arduino loop - copy sound to out 
void loop() {
   instrument.pluck(1.0 );
  instrument.noteOn(190, 0.9);
  
       instrument.pluck(0.7);
       delay(100);
  instrument.noteOn(3490, 0.7);
  delay(100);
  copier.copy();
    instrument.pluck(0.1);
  instrument.noteOn(1200, 0.9);
    
       instrument.pluck(0.7);
    delay(100);
  instrument.noteOn(890, 0.7);

  copier.copy();
 

       
    
 
}`

**Comment by SnazzyBlanket** [comment]
I attempted to use similar code but with 
I2SStream out;

this gave me errors and did not work

**Comment by SnazzyBlanket** [comment]
i think part of my confusion comes because there are examples in the audio tools and in the audio drivers folder. I am not sure WHICH one has been redacted etc. there are different examples on the wiki page for each one and they explain different ways to call the I2s driver.

so the more I read the documents the more I am confused about which is the proper way for the RP2040 to call the driver

I have added #define RP2040_HOWER to my code as well

**Comment by SnazzyBlanket** [comment]
thanks for any clarification on how to call the I2s driver from the pico (using the pinout for the pcm5102 which you specified on your page)

**Comment by SnazzyBlanket** [comment]
; PlatformIO Project Configuration File
;
;   Build options: build flags, source filter
;   Upload options: custom upload port, speed and extra flags
;   Library options: dependencies, extra library storages
;   Advanced options: extra scripting
;
; Please visit documentation for the other options and examples
; https://docs.platformio.org/page/projectconf.html
here are my platform IO settings as well


[env:pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower
upload_protocol = cmsis-dap
lib_deps = 
	khoih-prog/RP2040_PWM@^1.7.0
	bodmer/TFT_eSPI@^2.5.43
platform_packages =
    maxgerhardt/framework-arduinopico@https://github.com/earlephilhower/arduino-pico.git

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/wiki/Supported-Architectures

The RP2040 has no DAC, so you need to use an external DAC via I2S or use PWM. I also recommend to work with Earl Phil Howers Arduino Core. In Arduino this should just work. I never tried this for the RP2040 on PlatformIO, so you are on your own...

**Reply by SnazzyBlanket** [reply]
since you have never tried it--once i get it working i will post what i had to do to get it working here. i was finally able to get the code to compile

**Comment by SnazzyBlanket** [comment]
yes sorry i thought. mentionedi hooked up the dac you recommmended the pcm5102

**Comment by pschatzmann** [comment]
https://github.com/pschatzmann/arduino-audio-tools/wiki/Working-with-PlatformIO

**Comment by SnazzyBlanket** [comment]
I am having no issue getting other pico code to work in platform Io and I was also using platform IO for the ESP32 code I was using yesterday. The code is loading fine etc. which is what makes me wonder if my code itself is suspect

is this the correct way (if i have the Dac connected) to call the i2s Dac ?

typedef int16_t sound_t;                                   // sound will be represented as int16_t (with 2 bytes)
AudioInfo info(44100,1, 16); 
Sitar instrument(10);
STKStream<Instrmnt> in(instrument);
I2SStream out;
StreamCopy copier(out,in);                             // copies sound into i2s

**Comment by SnazzyBlanket** [comment]
I will try the code in arduino as well but just wanted to make sure I was not defining the i2s stream incorrectly for pico

**Reply by pschatzmann** [reply]
I suggest too use a discussion in audio-tools, since this project has nothing to do with what you are doing here
